pandas/core/internals/construction.py

```diff
@@ -687,10 +690,7 @@ def reorder_arrays(
                     arr = arrays[k]
                 new_arrays[i] = arr
 
-            # Incompatible types in assignment (expression has type
-            # "List[Union[ExtensionArray, ndarray[Any, Any], None]]", variable
-            # has type "List[Union[ExtensionArray, ndarray[Any, Any]]]")
-            arrays = new_arrays  # type: ignore[assignment]
+            arrays = cast(list[ArrayLike], new_arrays)
```

Comment:
is it possible to refactor instead of casting? it might be possible to be more precise here

something like
````python
            new_arrays: list[ArrayLike] = []
            indexer = arr_columns.get_indexer(columns)
            for i, k in enumerate(indexer):
                if k == -1:
                    # by convention default is all-NaN object dtype
                    arr = np.empty(length, dtype=object)
                    arr.fill(np.nan)
                else:
                    arr = arrays[k]
                new_arrays.append(arr)
```
then the cast might not be necessary? haven't tried running this (currently on a train :train:), but hoping this might work

Reply:
Thank you, I did as you suggested. It works very well.